### PR TITLE
Add CRUD support for registered limits

### PR DESCRIPTION
**Pull request by konan-abhi**
This PR will add support to Update, Delete, List and Get particular registered limit.

### CODE REVIEW

File: modules/openstack/limits.go [hunk 1]
```diff
@@ -151,3 +161,73 @@ func (o *OpenStack) CreateRegisteredLimit(
...
+
+//GetRegisteredLimit - Get existing registered limit by ID
+func (o *OpenStack) GetRegisteredLimit(
+	log logr.Logger,
+	registeredLimitID string,
+) (RegisteredLimit, error) {
+    log.Info(fmt.Sprintf("Fetching registered limit %s", registeredLimitID))
+	registered_limit, err := registeredlimits.Get(o.osclient, registeredLimitID).Extract()
+    if err != nil {
+        return _, err
```
Comment (fmount) [inline modules/openstack/limits.go:186]: `Error: vet: ./limits.go:186:16: cannot use _ as value or type`

File: modules/openstack/limits.go [hunk 2]
```diff
@@ -151,3 +161,73 @@ func (o *OpenStack) CreateRegisteredLimit(
...
+
+	log.Info(fmt.Sprintf("Fetching registered limit %s", resourceName))
+	allPages, err := registeredlimits.List(o.osclient, listOpts).AllPages()
+	if err != nil {
+		return _, err
+	}
+
+	allLimits, err := limits.ExtractRegisteredLimits(allPages)
+	if err != nil {
+		return _, err
```
Comment (fmount) [inline modules/openstack/limits.go:208]: ditto

File: modules/openstack/limits.go [hunk 3]
```diff
@@ -151,3 +161,73 @@ func (o *OpenStack) CreateRegisteredLimit(
...
+	serviceID string,
+) ([]RegisteredLimit, error) {
+    listOpts := registeredlimits.ListOpts{
+		ServiceID: serviceID,
+	}
+
+	log.Info(fmt.Sprintf("Fetching registered limit for service %s", serviceID))
+	allPages, err := registeredlimits.List(o.osclient, listOpts).AllPages()
+	if err != nil {
+		return _, err
```
Comment (fmount) [inline modules/openstack/limits.go:225]: ditto

File: modules/openstack/limits.go (hunk 1 above)
Comment (fmount) [reply modules/openstack/limits.go:186]: Should we change `RegisteredLimit` into something like `*registeredlimits.RegisteredLimit` and return it from the function? Just trying to figure out how to properly change this function.
We could have something like:

```
// GetRegisteredLimit - Get existing registered limit by ID
func (o *OpenStack) GetRegisteredLimit(
	log logr.Logger,
	registeredLimitID string,
) (*registeredlimits.RegisteredLimit, error) {
	log.Info(fmt.Sprintf("Fetching registered limit %s", registeredLimitID))
	registered_limit, err := registeredlimits.Get(o.osclient, registeredLimitID).Extract()
	if err != nil {
		return nil, err
	}
	return registered_limit, nil
}
```
but I have to try the whole flow and double check it works as expected.
Anyway, no need to use "_" in a return statement, we can return "nil" on that field.

Comment (fmount) [reply modules/openstack/limits.go:186]: Same comment applies to the other functions

**Review by fmount — CHANGES_REQUESTED** [review]

File: modules/openstack/limits.go [hunk 4]
```diff
@@ -151,3 +161,73 @@ func (o *OpenStack) CreateRegisteredLimit(
...
+		ResourceName: resourceName,
+	}
+
+	log.Info(fmt.Sprintf("Fetching registered limit %s", resourceName))
+	allPages, err := registeredlimits.List(o.osclient, listOpts).AllPages()
+	if err != nil {
+		return _, err
+	}
+
+	allLimits, err := limits.ExtractRegisteredLimits(allPages)
```
Comment (fmount) [inline modules/openstack/limits.go:206]: that should be `registeredlimits.ExtractRegisteredLimits(allPages)`

File: modules/openstack/limits.go [hunk 5]
```diff
@@ -151,3 +161,73 @@ func (o *OpenStack) CreateRegisteredLimit(
...
+        return _, err
+    }
+    return registered_limit, nil
+}
+
+//ListRegisteredLimitsByResourceName - List all registered limits filtered by resource name
+func (o *OpenStack) ListRegisteredLimitsByResourceName(
+	log logr.Logger,
+	resourceName string,
+) ([]RegisteredLimit, error) {
```
Comment (fmount) [inline modules/openstack/limits.go:195]: this should be `[]registeredlimits.RegisteredLimit`)

File: modules/openstack/limits.go [hunk 6]
```diff
@@ -151,3 +161,73 @@ func (o *OpenStack) CreateRegisteredLimit(
...
+    listOpts := registeredlimits.ListOpts{
+		ResourceName: resourceName,
+	}
+
+	log.Info(fmt.Sprintf("Fetching registered limit %s", resourceName))
+	allPages, err := registeredlimits.List(o.osclient, listOpts).AllPages()
+	if err != nil {
+		return _, err
```
Comment (fmount) [inline modules/openstack/limits.go:203]: return nil, err

**Review by fmount — CHANGES_REQUESTED** [review]

File: modules/openstack/limits.go (hunk 3 above)
Comment (konan-abhi) [reply modules/openstack/limits.go:225]: Done

File: modules/openstack/limits.go [hunk 7]
```diff
@@ -151,3 +161,73 @@ func (o *OpenStack) CreateRegisteredLimit(
...
+
+	log.Info(fmt.Sprintf("Fetching registered limit for service %s", serviceID))
+	allPages, err := registeredlimits.List(o.osclient, listOpts).AllPages()
+	if err != nil {
+		return _, err
+	}
+
+	allLimits, err := limits.ExtractRegisteredLimits(allPages)
+	if err != nil {
+		return _, err
```
[the remark this answers is not in the document]
Comment (konan-abhi) [reply modules/openstack/limits.go:230]: Done

File: modules/openstack/limits.go (hunk 4 above)
Comment (konan-abhi) [reply modules/openstack/limits.go:206]: Done

**Comment by fmount** [comment]
Thanks @konan-abhi for the update here! looks good, I'll test it locally and get back to you!

**Comment by fmount** [comment]
ok I think we're good here. Thanks @konan-abhi this last change works as expected.

```
> oc apply -f glance_v1beta1_glance.file_quota.yaml
glance.glance.openstack.org/glance created
(venv)

> openstack registered limit list
+----------------------------------+----------------------------------+-----------------------+---------------+------------------------------------------+-----------+
| ID                               | Service ID                       | Resource Name         | Default Limit | Description                              | Region ID |
+----------------------------------+----------------------------------+-----------------------+---------------+------------------------------------------+-----------+
| cbc0a91221de4fb48a563f87505e85bc | 9ba4a7211b184ded8ff5f8e9e59e29ac | image_count_uploading |           150 | default limit for  image_count_uploading | regionOne |
| af31dcc228f44bd8ab7e88cda20795bf | 9ba4a7211b184ded8ff5f8e9e59e29ac | image_count_total     |          1001 | default limit for  image_count_total     | regionOne |
| 1081eee6d3d146db85b2e272bd95da5a | 9ba4a7211b184ded8ff5f8e9e59e29ac | image_stage_total     |           100 | default limit for  image_stage_total     | regionOne |
| f5b923d2394a45f4b64de741d51a3e6d | 9ba4a7211b184ded8ff5f8e9e59e29ac | image_size_total      |           100 | default limit for  image_size_total      | regionOne |
+----------------------------------+----------------------------------+-----------------------+---------------+------------------------------------------+-----------+
(venv)
> oc delete -f glance_v1beta1_glance.file_quota.yaml
glance.glance.openstack.org "glance" deleted
(venv)
> openstack registered limit list

```

**Review by fmount — APPROVED** [review]

**Pull request merged**